### PR TITLE
Add container schema support to inventory-app

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Inventory-app has some code that is intended to be app-independent logic for using SharedTree to author react apps.

Turns out its actually possible to author a function which we could expose as `@public` that produces a `DataObjectClass` that can be used with the public container schema APIs: this PR does this, allowing tree schema to be baked into container schema in a type safe way, including handling of documents with incompatible tree stored schema.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: examples/data-objects/inventory-app/src/reactSharedTreeView.tsx [hunk 1]
```diff
@@ -73,32 +141,10 @@ export abstract class TreeDataObject<
 		if (this.#tree === undefined) throw new Error(this.getUninitializedErrorString("tree"));
 	}
 
-	public abstract key: string;
+	public abstract readonly key: string;
```
Comment (Josmithr) [inline examples/data-objects/inventory-app/src/reactSharedTreeView.tsx:144]: Can we re-document the abstract members here?

File: examples/data-objects/inventory-app/src/inventoryList.ts [hunk 2]
```diff
@@ -4,21 +4,42 @@
...
 import { type Inventory, treeConfiguration } from "./schema.js";
 
 /**
- * @internal
+ * For use with high level service-client container API's like AzureClient and OdspClient.
+ * @privateRemarks
+ * TODO: There is no service implementation agnostic client abstraction that can be referred to here (ex: shared by AzureClient and OdspClient).
+ * This makes documenting compatibility with that implicit common API difficult.
+ * It also makes writing service agnostic code at that abstraction level harder.
+ * This should be fixed.
```
Comment (Josmithr) [inline examples/data-objects/inventory-app/src/inventoryList.ts:18]: \+1

File: examples/data-objects/inventory-app/src/inventoryList.ts [hunk 3]
```diff
@@ -4,21 +4,42 @@
...
+ *
+ * TODO:
+ * Writing an app at this abstraction level currently requires a lot of boilerplate which also requires extra dependencies.
+ * Since `@fluid-example/example-utils` doesn't provide that boilerplate and neither do the public packages, there isn't a concise way to actually use this container in this example.
+ * This should be fixed.
+ *
+ * TODO:
+ * The commonly used boilerplate for setting up a ContainerSchema based application configures the dev-tools, which would be great to include in this example,
```
Comment (Josmithr) [inline examples/data-objects/inventory-app/src/inventoryList.ts:26]: I'm not sure I understand this comment. You should be able to create the dev-tools by importing:
```typescript
import { initializeDevtools } from "@fluidframework/devtools";
```

Comment (Josmithr) [reply examples/data-objects/inventory-app/src/inventoryList.ts:26]: Or are you suggesting the data object scheme here is incompatible. That may be the case, and if so should be investigated.

Comment (CraigMacomber) [reply examples/data-objects/inventory-app/src/inventoryList.ts:26]: if you do that fluid layer check objects. There is apparently work planned to fix that (make dev tools not depend on exmaples)

Comment (CraigMacomber) [reply examples/data-objects/inventory-app/src/inventoryList.ts:26]: More generally I'd like to include the boilerplate like https://github.com/microsoft/FluidExamples/blob/625b8fdb7d2c8638cefba293e2fa8520ae8c5e84/brainstorm/src/index.tsx and everything from https://github.com/microsoft/FluidExamples/tree/625b8fdb7d2c8638cefba293e2fa8520ae8c5e84/brainstorm/src/infr

Comment (Josmithr) [reply examples/data-objects/inventory-app/src/inventoryList.ts:26]: Ugh. Okay, I understand. I _think_ you can add exceptions to the layer checker, (the dependency from devtools is a dev dependency only, so adding the import shouldn't cause any real issues). But if not, the work to clean up that dependency is scheduled for this sprint at least.

Comment (CraigMacomber) [reply examples/data-objects/inventory-app/src/inventoryList.ts:26]: Eventually (once this is split into its own package so others can use it), resolving this limit will be much more valuable, but for now this can just wait until the issue is resolved.

File: packages/framework/fluid-static/api-report/fluid-static.api.md [hunk 4]
```diff
@@ -44,7 +44,7 @@ export function createServiceAudience<TMember extends IMember = IMember>(props:
 // @public
 export type DataObjectClass<T extends IFluidLoadable = IFluidLoadable> = {
     readonly factory: {
-        IFluidDataStoreFactory: DataObjectClass<T>["factory"];
+        readonly IFluidDataStoreFactory: DataObjectClass<T>["factory"];
```
Comment (Josmithr) [inline packages/framework/fluid-static/api-report/fluid-static.api.md:47]: This change seems good on its own. Should we split this into a separate PR?

File: examples/data-objects/inventory-app/src/inventoryList.ts [hunk 5]
```diff
@@ -4,21 +4,42 @@
...
+ * but can't be included due to dependency layering issues.
  */
+export const containerSchema = {
```
Comment (ChumpChief) [inline examples/data-objects/inventory-app/src/inventoryList.ts:29]: Where does this get used?  Am I missing it?

Comment (CraigMacomber) [reply examples/data-objects/inventory-app/src/inventoryList.ts:29]: it is not currently used since using it is totally impractical due to needing a ton of boilerplate and deps. However this is still an example of how we could make tree dataobjects compatible with container schema, which I think is important

Comment (CraigMacomber) [reply examples/data-objects/inventory-app/src/inventoryList.ts:29]: If the above TODO's were addressed, we could use this here.

File: packages/framework/fluid-static/api-report/fluid-static.api.md (hunk 4 above)
Comment (CraigMacomber) [reply packages/framework/fluid-static/api-report/fluid-static.api.md]: DOne.

File: examples/data-objects/inventory-app/src/inventoryList.ts (hunk 5 above)
Comment (CraigMacomber) [reply examples/data-objects/inventory-app/src/inventoryList.ts]: I have moved this code into a unit test with different docs to clarify.

File: examples/data-objects/inventory-app/tests/main.test.ts [hunk 6]
```diff
@@ -21,4 +24,34 @@ describe.skip("Main", () => {
 	it("The page loads", async () => {
 		// await expect(page).toClick("button", { text: "Roll" });
 	});
+
+	it("works with container schema", async () => {
+		// await expect(page).toClick("button", { text: "Roll" });
```
Comment (Josmithr) [inline examples/data-objects/inventory-app/tests/main.test.ts:29]: Is this no-op test desired? Neither this nor the test above actually validate anything (beyond that the page doesn't throw on load). Do we have anything tracking getting these tests working?

**Review by Josmithr — APPROVED** [review]
Left a comment in the tests, otherwise seems good to me.

File: examples/data-objects/inventory-app/tests/main.test.ts (hunk 6 above)
Comment (CraigMacomber) [reply examples/data-objects/inventory-app/tests/main.test.ts:29]: oops, I added that when deciding to make a separate suite since this one is skipped. I'll remove it.

I can't run any of our jest tests locally (since they all give file watching errors) despite the entire test suite being skipped and the test bodies being commented out. Somone should care about that someday. I don't know if we are tracking the very bad state of many of our examples having no tests.

Comment (CraigMacomber) [reply examples/data-objects/inventory-app/tests/main.test.ts]: Fixed (well I delated the extra test, we still have skipped commented out tests with broken IntelliSense which I can't run locally)
